doc/axiom.md

```diff
@@ -66,51 +68,121 @@ The characters after the underscore can be written as any word character; period
 
-In Flint, string literals can contain any characters **except** unescaped line breaks, carriage returns, backslashes, or double quotes; this is in order to prevent SPARQL injection attacks. (Therefore strings like `"cat: \"meow\"` and `"foo\nbar"` are not allowed.) Strings are not transformed during SPARQL translation.
+String literals can contain any characters **except** unescaped line breaks, carriage returns, backslashes, or double quotes; this is in order to prevent SPARQL injection attacks. (Therefore strings like `"cat: \"meow\"` and `"foo\nbar"` are not allowed.) Strings are not transformed during SPARQL translation.
 
-### Language-tagged strings
+### Language Maps
 
 Examples: `{:en "Hello World!"}`, `{:zh "你好世界"}`
 
-In Flint, strings with language tags are represented by a map between **one** language tag keyword and the string literal (which follows the same restrictions as simple strings).
+Strings with language tags are represented by a map between **one** language tag keyword and the string literal (which follows the same restrictions as simple strings). Note that this is the only literal that cannot have a datatype IRI appended, even if `:force-iris?` is `true`.
 
-### dateTime timestamps
+### Timestamps
 
 Examples: `#inst "2022-01-01T10:10:10Z"`
 
-In Flint, dateTime timestamps are any values for which `inst?` is `true`. During SPARQL translation, an IRI denoting the datatype is added as a suffix (since dateTime timestamps, unlike other literals, do not have a non-suffixed representation in SPARQL):
-```sparql
-"2022-01-01T10:10:10Z"^^<http://www.w3.org/2001/XMLSchema#dateTime>
+Timestamps are any values for which `inst?` is `true`, i.e. an instance of one of the following:
+- `java.time.Instant` (Clojure)
+- `java.util.Date` (Clojure)
+- `java.sql.Date` (Clojure)
+- `java.sql.Time` (Clojure)
+- `js/Date` (ClojureScript)
+
+As mentioned above, timestamps will be stringified (in the Clojure case, they will be stringified as `java.time.Instant` instances), and will have the `xsd:dateTime` IRI appended regardless of the value of `:force-iris?`.
+
+### Custom Literals
+
+A user can implement a custom literal by extending `com.yetanalytics.flint.axiom.protocol/Literal` and defining the following:
+- `-valid-literal?` to validate the value.
+- `-format-literal` to format the entire string.
+- `-format-literal-strval`, `-format-literal-lang-tag`, and `-format-literal-url` to format individual aspects of the literal value.
+
+Note that `-format-literal` and `-format-literal-url` also accept an `opts` map for optional arguments that affect formatting. Currently implement literals accept the args `:force-iri?` (to force-append datatype IRIs) and `:iri-prefix-m` (a map from base IRI strings to prefixes to shorten datatype IRIs).
+
+Here is an example of an implementation of a `Rational` literal (inspired by an example given by the [Apache Jena documentation](https://jena.apache.org/documentation/notes/typed-literals.html)):
```

Comment:
The example is great, I like that it tracks Jena, but could possibly be a little more Clojure-y with minimal effort using [clojure.lang.Ratio](https://clojure.org/reference/data_structures#_ratio). Or I don't know, maybe you want to build that one in.

for reference:

```clojure
user=> 1/2
1/2
user=> (type 1/2)
clojure.lang.Ratio
user=> (.numerator 1/2)
1
user=> (.denominator 1/2)
2
```

Reply:
One advantage my custom `Rational` record has is that it demonstrates a custom `-valid-literal?` function implementation (especially with the division by zero check). Maybe we can have both examples?

Reply:
Actually it was very good that you suggested this, because I was able to uncover a bug with how the axiom specs were defined. When you have a spec be directly defined on a protocol fn:
```clojure
(def literal-spec
  p/-valid-literal?)
```
it will keep that definition even if you extend the protocol to additional types, causing validation to fail. Instead you'll have to resolve the protocol function at runtime, as so:
```clojure
(def literal-spec
  #(p/-valid-literal? %))
```

Reply:
Anyways fixed that bug in https://github.com/yetanalytics/flint/pull/27/commits/43a98d8e51eda6218a5d73f92a393bd4e862437d, and added the `clojure.lang.Ratio` example in https://github.com/yetanalytics/flint/pull/27/commits/0ca7b0dbf4d756bb14e0a65c717e691c2993abd0

Reply:
> I assume we're leaving clojure.lang.Ratio as an example because cljs doesn't support it?

You mean the custom `Rational` record? I guess so.

Reply:
> You mean the custom Rational record? I guess so.

No, I meant extending the literal protocol for `clojure.lang.Ratio` specifically, in `impl.cljc` (rather than just in an example). I do note that you added things like BigInt that don't exist (to my knowledge) in cljs though, so maybe lopsided-ness isn't a concern.

Reply:
Yeah `clojure.lang.Ratio` doesn't have a counterpart in cljs: https://clojurescript.org/about/differences#_the_reader
> ClojureScript currently only supports integer and floating point literals that map to JavaScript primitives
> - Ratio, BigDecimal, and BigInteger literals are currently not supported

Also if you want to see an example of both the record and `extend-type` in action, I have them as test cases: https://github.com/yetanalytics/flint/blob/protocols/src/test/com/yetanalytics/flint/axiom_test.cljc#L396

Reply:
> Yeah clojure.lang.Ratio doesn't have a counterpart in cljs

Right, I mentioned this. I was asking if that is _why_ you decided not to extend it by default. It's seldom used and since it is trivial for the lib consumer to add, I think it is fine the way you have it here.

Reply:
The real reason is that it's not a "basic" type and doesn't have a corresponding datatype URL (unlike ints with `xsd:integer` or booleans with `xsd:boolean`). Which goes along with what you said - it's seldom used and would be more suitable for the user to add it post-hoc.